### PR TITLE
[extractor/tvp] switch vod completely to vod api

### LINKED ISSUE #6987: [tvp] "Video requires payment and log-in, but log-in is not implemented"
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting that a **supported** site is broken
- [X] I've verified that I'm running yt-dlp version **2023.03.04** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

Canada

### Provide a description that is worded well enough to be understood

I've been downloading episodes from website: Vod.tvp.pl / https://vod.tvp.pl/seriale,18/polowanie-na-cmy-odcinki,390116

I've been able to get every episode with no issues what so ever, with the exception of Episode #7. https://vod.tvp.pl/seriale,18/polowanie-na-cmy-odcinki,390116/odcinek-7,S01E07,398869

[tvp:vod] Extracting URL: https://vod.tvp.pl/seriale,18/polowanie-na-cmy-odcinki,390116/odcinek-7,S01E07,398869
[tvp:vod] 398869: Downloading JSON metadata
[tvp:embed] Extracting URL: tvp:69230468
[tvp:embed] 69230468: Downloading webpage
ERROR: [tvp:embed] 69230468: Video requires payment and log-in, but log-in is not implemented; please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U

(Same error with -vU)

The video works in browser, I've created an account and tried --cookies-from-browser chrome with no luck. 

I've updated twice hoping for a resolution - video has been available since April 30th, 2023. 

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['-vU', 'https://vod.tvp.pl/seriale,18/polowanie-na-cmy-odcinki,390116/odcinek-7,S01E07,398869']
[debug] Encodings: locale cp1252, fs utf-8, pref cp1252, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version nightly@2023.05.03.204705 [2f07c4c1d] (win_exe)
[debug] Python 3.8.10 (CPython AMD64 64bit) - Windows-10-10.0.19044-SP0 (OpenSSL 1.1.1k  25 Mar 2021)
[debug] exe versions: none
[debug] Optional libraries: Cryptodome-3.17, brotli-1.0.9, certifi-2022.12.07, mutagen-1.46.0, sqlite3-2.6.0, websockets-11.0.2
[debug] Proxy map: {}
[debug] Loaded 1813 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp-nightly-builds/releases/latest
Available version: nightly@2023.05.03.204705, Current version: nightly@2023.05.03.204705
Current Build Hash: b73b1cb4c64dcbd61a084ba9910dca9837cedb7b450c75605ec5d197cdc8ad01
yt-dlp is up to date (nightly@2023.05.03.204705)
[tvp:vod] Extracting URL: https://vod.tvp.pl/seriale,18/polowanie-na-cmy-odcinki,390116/odcinek-7,S01E07,398869
[tvp:vod] 398869: Downloading JSON metadata
[tvp:embed] Extracting URL: tvp:69230468
[tvp:embed] 69230468: Downloading webpage
ERROR: [tvp:embed] 69230468: Video requires payment and log-in, but log-in is not implemented; please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
  File "yt_dlp\extractor\common.py", line 694, in extract
  File "yt_dlp\extractor\tvp.py", line 392, in _real_extract
```

### PR DESCRIPTION
**Pull request by selfisekai**
fixes https://github.com/yt-dlp/yt-dlp/issues/6987

**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

switches extraction of vod completely to the new vod apis

Fixes #6987

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

File: yt_dlp/extractor/tvp.py [hunk 1]
```diff
@@ -482,21 +482,32 @@ def _real_extract(self, url):
...
-        return {
-            '_type': 'url',
-            'url': 'tvp:' + video['externalUid'],
-            'ie_key': TVPEmbedIE.ie_key(),
+    def _parse_video(self, video, with_url=True):
+        info_dict = {
+            'id': str(video['id']),
             'title': video.get('title'),
-            'description': traverse_obj(video, ('lead', 'description')),
+            'description': clean_html(video.get('lead') or video.get('description')),
```
Comment (pukkandan) [inline yt_dlp/extractor/tvp.py:494]: Old code was looking at `video.load.description`. Is this change intentional? Was the old code wrong? Don't tests change accordingly?

```suggestion
            'description': traverse_obj(video, 'lead', 'description', expected_type=clean_html),
```

File: yt_dlp/extractor/tvp.py [hunk 2]
```diff
@@ -482,21 +482,32 @@ def _real_extract(self, url):
...
             'age_limit': int_or_none(video.get('rating')),
             'duration': int_or_none(video.get('duration')),
+            'episode_number': int_or_none(video.get('number')),
+            'series': traverse_obj(video, ('season', 'serial', 'title')),
+            'thumbnails': [{'url': u}
+                           for u in traverse_obj(video,
+                           ('images', ..., ..., 'url'), expected_type=url_or_none)
+                           if u is not None],
         }
```
Comment (pukkandan) [inline yt_dlp/extractor/tvp.py:503]: Actually, this whole thing can be simplified with new traverse_obj features.
```suggestion
        info_dict = traverse_obj(video, {
            'id': ('id', {str_or_none}),
            'title': 'title',
            'age_limit': ('rating', {int_or_none}),
            'duration': ('duration', {int_or_none}),
            'episode_number': ('number', {int_or_none}),
            'series': ('season', 'serial', 'title', {str}),
            'thumbnails': ('images', ..., ..., {'url': ('url', {url_or_none})}),
        })
        info_dict['description'] = traverse_obj(video, 'lead', 'description', expected_type=clean_html)
```

File: yt_dlp/extractor/tvp.py [hunk 3]
```diff
@@ -447,7 +447,7 @@ def _real_extract(self, url):
         duration = try_get(info, lambda x: x['duration'], int) if not is_live else None
 
         subtitles = {}
-        for sub in content.get('subtitles') or []:
+        for sub in content.get('subtitles') or ():
```
Comment (pukkandan) [inline yt_dlp/extractor/tvp.py:450]:
```suggestion
        for sub in content.get('subtitles') or []:
```

unnecessary

File: yt_dlp/extractor/tvp.py [hunk 4]
```diff
@@ -506,37 +517,73 @@ class TVPVODVideoIE(TVPVODBaseIE):
...
+            lang = sub['language']
+            subtitles.setdefault(lang, []).append({
+                'url': sub['url'],
+                'ext': 'ttml',
+            })
+
+        info_dict['formats'] = formats
+        info_dict['subtitles'] = subtitles
+
+        return info_dict
```
Comment (pukkandan) [inline yt_dlp/extractor/tvp.py:586]:
```suggestion
        info_dict['formats'] = []
        for manifest_url in traverse_obj(playlist, ('sources', 'HLS', ..., 'src')):
            info_dict['formats'].extend(self._extract_m3u8_formats(manifest_url, video_id, fatal=False))
        for manifest_url in traverse_obj(playlist, ('sources', 'DASH', ..., 'src')):
            info_dict['formats'].extend(self._extract_mpd_formats(manifest_url, video_id, fatal=False))

        info_dict['subtitles'] = {}
        for sub in playlist.get('subtitles') or []:
            info_dict['subtitles'].setdefault(sub.get('language') or 'und', []).append({
                'url': sub['url'],
                'ext': 'ttml',
            })

        return info_dict
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/tvp.py (hunk 1 above)
Comment (selfisekai) [reply yt_dlp/extractor/tvp.py:494]: it was supposed to be lead or description, tests apparently had both so far and it worked

File: yt_dlp/extractor/tvp.py (hunk 2 above)
Comment (Grub4K) [reply yt_dlp/extractor/tvp.py]: This was the reason I had behind wanting to discard empty string.

Alternatively we can finally introduce `truthy_or_none` and do
```py
description = traverse_obj(video, (('lead', 'description'), {clean_html}, {truthy_or_none}), get_all=False)
```

`.get()` can throw if input data is not a dict.

**Comment by selfisekai** [comment]
also added error handling now, for stuff that is actually paid

on master:
```
[tvp:vod] Extracting URL: https://vod.tvp.pl/seriale,18/korona-krolow-jagiellonowie-odcinki,292227/odcinek-53,S01E53,399444
[tvp:vod] 399444: Downloading JSON metadata
[tvp:embed] Extracting URL: tvp:69359478
[tvp:embed] 69359478: Downloading webpage
ERROR: [tvp:embed] 69359478: Video requires payment and log-in, but log-in is not implemented; please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
```

before last commit:
```
[tvp:vod] Extracting URL: https://vod.tvp.pl/seriale,18/korona-krolow-jagiellonowie-odcinki,292227/odcinek-53,S01E53,399444
[tvp:vod] 399444: Downloading JSON metadata
[tvp:vod] 399444: Downloading JSON metadata
ERROR: [tvp:vod] 399444: Unable to download JSON metadata: HTTP Error 403: Forbidden (caused by <HTTPError 403: 'Forbidden'>); please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
```

after:
```
[tvp:vod] Extracting URL: https://vod.tvp.pl/seriale,18/korona-krolow-jagiellonowie-odcinki,292227/odcinek-53,S01E53,399444
[tvp:vod] 399444: Downloading JSON metadata
[tvp:vod] 399444: Downloading JSON metadata
ERROR: [tvp:vod] 399444: Response from Woronicza: ITEM_NOT_PAID; please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
```

File: yt_dlp/extractor/tvp.py (hunk 2 above)
Comment (pukkandan) [reply yt_dlp/extractor/tvp.py]: > `.get()` can throw if input data is not a dict.

That's not an issue here. Ig, we can go without traverse_obj for the time being. But still `dict_get` is better than `.get` imo.

```py
info_dict['description'] = clean_html(dict_get(video, 'lead', 'description'))
```

File: yt_dlp/extractor/tvp.py [hunk 5]
```diff
@@ -483,9 +483,16 @@ class TVPVODBaseIE(InfoExtractor):
     _API_BASE_URL = 'https://vod.tvp.pl/api/products'
 
     def _call_api(self, resource, video_id, query={}, **kwargs):
-        return self._download_json(
+        is_valid = lambda x: x >= 200 and x < 300
+        is_handled_error = lambda x: x >= 400 and x < 500
```
Comment (pukkandan) [inline yt_dlp/extractor/tvp.py:487]:
```suggestion
        is_valid = lambda x: 200 <= x < 300
        is_handled_error = lambda x: is_valid(x) or 400 <= x < 500
```
is_handled_error is never used by itself

File: yt_dlp/extractor/tvp.py [hunk 6]
```diff
@@ -483,9 +483,16 @@ class TVPVODBaseIE(InfoExtractor):
...
+        document, urlh = self._download_json_handle(
             f'{self._API_BASE_URL}/{resource}', video_id,
-            query={'lang': 'pl', 'platform': 'BROWSER', **query}, **kwargs)
+            query={'lang': 'pl', 'platform': 'BROWSER', **query},
+            expected_status=lambda x: is_valid(x) or is_handled_error(x), **kwargs)
```
Comment (pukkandan) [inline yt_dlp/extractor/tvp.py:491]:
```suggestion
            expected_status=is_handled_error, **kwargs)
```

File: yt_dlp/extractor/tvp.py [hunk 7]
```diff
@@ -483,9 +483,16 @@ class TVPVODBaseIE(InfoExtractor):
...
+        if is_valid(urlh.status):
+            return document
+        error_code = document.get('code')
+        raise ExtractorError(f'Response from Woronicza: {error_code}')
```
Comment (pukkandan) [inline yt_dlp/extractor/tvp.py:495]:
```suggestion
        raise ExtractorError(f'Woronicza said: {document.get("code")}')
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/tvp.py (hunk 2 above)
Comment (bashonly) [reply yt_dlp/extractor/tvp.py]: I think it'd need to be
```py
info_dict['description'] = clean_html(dict_get(video, ('lead', 'description')))
```

Comment (selfisekai) [reply yt_dlp/extractor/tvp.py]: changed

**Comment by Krzysiu** [comment]
Since the update default title seems to be wrong for series. It used to be `%series_name% %tile%` which equals to "odc." ("ep.")  and number `[%id%]`, so for example for series foo, ep. 50, it was "Foo, odc. 50 [123456]". Now it's just "odc. 50 [123456].

JSON shows (example series, don't blame me) {"id": "579561", "title": "odc. 120", "age_limit": 12, "duration": 2446, "episode_number": 120, "series": "Akacjowa 38", "thumbnails": [{"url": "http://s.tvp.pl/images/3/2/e/uid_32e96c96fd7b495b8453d3d24a92f40b_width_315_play_0_pos_0_gs_0_height_177.jpg", "id": "0"}]

I think it would be better to prepend `series` value to `title`, if `series` exists.
